stream.go

```diff
@@ -0,0 +1,185 @@
+// received) for the stream rid and may change
+func (s *TrackRTPStream) SSRC() uint32 {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.ssrc
+}
+
+// Codec gets the Codec of the stream. If this is a rid based stream
+// the codec may be nil until a packet is received.
+func (s *TrackRTPStream) Codec() *RTPCodec {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.codec
+}
+
+// Packetizer gets the Packetizer of the stream
+func (s *TrackRTPStream) Packetizer() rtp.Packetizer {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.packetizer
+}
+
+// Read reads data from the stream. If this is a local stream this will error
+func (s *TrackRTPStream) Read(b []byte) (n int, err error) {
+	return s.track.read(b, s.rid)
+}
+
+// ReadRTP is a convenience method that wraps Read and unmarshals for you
+func (s *TrackRTPStream) ReadRTP() (*rtp.Packet, error) {
+	b := make([]byte, receiveMTU)
+	i, err := s.Read(b)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &rtp.Packet{}
+	if err := r.Unmarshal(b[:i]); err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+// Write writes data to the stream. If this is a remote stream this will error
+func (s *TrackRTPStream) Write(b []byte) (n int, err error) {
+	packet := &rtp.Packet{}
+	err = packet.Unmarshal(b)
+	if err != nil {
+		return 0, err
+	}
+
+	err = s.WriteRTP(packet)
+	if err != nil {
+		return 0, err
+	}
+
+	return len(b), nil
+}
+
+// WriteSample packetizes and writes to the stream
+func (s *TrackRTPStream) WriteSample(sample media.Sample) error {
+	packets := s.packetizer.Packetize(sample.Data, sample.Samples)
+	for _, p := range packets {
+		err := s.WriteRTP(p)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// WriteRTP writes RTP packets to the stream
+func (s *TrackRTPStream) WriteRTP(p *rtp.Packet) error {
+	return s.track.WriteRTP(p)
+}
+
+// NewTrackRTPStream initializes a new RTPStream
+func NewTrackRTPStream(rid string, payloadType uint8, ssrc uint32, codec *RTPCodec) (*TrackRTPStream, error) {
+	if ssrc == 0 {
+		return nil, fmt.Errorf("SSRC supplied to NewStream() must be non-zero")
+	}
+
+	streamID := strconv.FormatUint(uint64(ssrc), 10)
+	// if rid is not empty use it as stream id
+	if rid != "" {
+		streamID = rid
+	}
+
+	packetizer := rtp.NewPacketizer(
+		rtpOutboundMTU,
+		payloadType,
+		ssrc,
+		codec.Payloader,
+		rtp.NewRandomSequencer(),
+		codec.ClockRate,
+	)
+
+	return &TrackRTPStream{
+		id:          streamID,
+		rid:         rid,
+		payloadType: payloadType,
+		ssrc:        ssrc,
+		codec:       codec,
+		packetizer:  packetizer,
+	}, nil
+}
+
+// determinePayloadType blocks and reads a single packet to determine the PayloadType for this Stream
+// this is useful if we are dealing with a remote stream and we can't announce it to the user until we know the payloadType
+func (s *TrackRTPStream) determinePayloadType() error {
+	r, err := s.ReadRTP()
+	if err != nil {
+		return err
+	}
+
+	s.mu.Lock()
+	s.payloadType = r.PayloadType
+	defer s.mu.Unlock()
```

Comment:
Unnecessary `defer`

Reply:
Just copied the current implementation in Track. I'll remove the defer.